File: py/rt/final/Sphere.py
from Hitable import Hitable, HitRecord
from Vec3 import Vec3
from Ray import Ray
from math import sqrt

class Sphere(Hitable):
	def __init__(self, center, radius, mat):
		self.center = center
		self.radius = radius
		self.mat = mat

	def hit(self, r, tMin, tMax):
		a = r.direction().lengthSquared()
		oc = r.origin() - self.center
		halfB = Vec3.dot(oc, r.direction())
		c = oc.lengthSquared() - self.radius*self.radius
		discriminant = halfB*halfB - a*c

		if discriminant > 0:
			root = sqrt(discriminant)

			t = (-halfB - root) / a
			if t < tMax and t > tMin:
				p = r.pointAtParameter(t)
				outwardNormal = (p - self.center) / self.radius
				return HitRecord(t,p,self.mat,r,outwardNormal)
			
			t = (-halfB + root) / a
			if t < tMax and t > tMin:
				p = r.pointAtParameter(t)
				outwardNormal = (p - self.center) / self.radius
				return HitRecord(t,p,self.mat,r,outwardNormal)
		
		return None		
		

File: cpp/OpenGL/22_Raytrace/Sphere.cpp
#include <cmath>

#include "Sphere.h"

Sphere::Sphere(const Vec3& center, const float radius, const std::shared_ptr<Material> materal) :
	center{center},
	radius{radius},
	material{materal}
{	
}

const std::optional<HitRecord> Sphere::hit(const Ray &r, const float tMin, const float tMax) const {
	
	const float a = r.direction().sqlength();
	const Vec3 oc{r.origin() - center};
	const float halfB = Vec3::dot(oc, r.direction());
	const float c = oc.sqlength() - radius*radius;
	const float discriminant = halfB*halfB - a*c;

	if (discriminant > 0) {
		const float root = sqrt(discriminant);

		const float t1 = (-halfB - root) / a;
		if (t1 < tMax && t1 > tMin) {
			const Vec3 p{r.pointAtParameter(t1)};
			const Vec3 outwardNormal{(p - center) / radius};
			return HitRecord{t1,p,outwardNormal,r,material};
		}
		
		const float t2 = (-halfB + root) / a;
		if (t2 < tMax && t2 > tMin) {
			const Vec3 p{r.pointAtParameter(t2)};
			const Vec3 outwardNormal{(p - center) / radius};
			return HitRecord{t2,p,outwardNormal,r,material};
		}
	}
	
	return {};
}